From d71a98c23c43a06650e90480e864c0e3da344437 Mon Sep 17 00:00:00 2001
From: Bo YU <tsu.yubo@gmail.com>
Date: Tue, 25 Apr 2023 03:50:43 +0200
Subject: [PATCH] arch: Add support for riscv32 CPU

[guillem@debian.org:
  - Adapt the test suite.
  - Add ABI checking support. ]

Closes: #1034756
Signed-off-by: Guillem Jover <guillem@debian.org>
---
 data/cputable                  | 1 +
 scripts/Dpkg/Shlibs/Objdump.pm | 9 +++++++++
 scripts/t/Dpkg_Arch.t          | 4 ++--
 3 files changed, 12 insertions(+), 2 deletions(-)

diff --git a/data/cputable b/data/cputable
index 7b1ee2c58..7c2c1c1c8 100644
--- a/data/cputable
+++ b/data/cputable
@@ -45,6 +45,7 @@ powerpc		powerpc		(powerpc|ppc)		32	big
 powerpcel	powerpcle	powerpcle		32	little
 ppc64		powerpc64	(powerpc|ppc)64		64	big
 ppc64el		powerpc64le	powerpc64le		64	little
+riscv32		riscv32		riscv32			32	little
 riscv64		riscv64		riscv64			64	little
 s390		s390		s390			32	big
 s390x		s390x		s390x			64	big
diff --git a/scripts/Dpkg/Shlibs/Objdump.pm b/scripts/Dpkg/Shlibs/Objdump.pm
index 5e27ec7f2..e30921b68 100644
--- a/scripts/Dpkg/Shlibs/Objdump.pm
+++ b/scripts/Dpkg/Shlibs/Objdump.pm
@@ -118,6 +118,7 @@ use constant {
     EM_XTENSA               => 94,
     EM_MICROBLAZE           => 189,
     EM_ARCV2                => 195,
+    EM_RISCV                => 243,
     EM_LOONGARCH            => 258,
     EM_AVR_OLD              => 0x1057,
     EM_OR1K_OLD             => 0x8472,
@@ -143,6 +144,13 @@ use constant {
 
     EF_IA64_ABI64           => 0x00000010,
 
+    EF_RISCV_FLOAT_ABI_SOFT     => 0x0000,
+    EF_RISCV_FLOAT_ABI_SINGLE   => 0x0002,
+    EF_RISCV_FLOAT_ABI_DOUBLE   => 0x0004,
+    EF_RISCV_FLOAT_ABI_QUAD     => 0x0006,
+    EF_RISCV_FLOAT_ABI_MASK     => 0x0006,
+    EF_RISCV_RVE                => 0x0008,
+
     EF_LOONGARCH_SOFT_FLOAT     => 0x00000001,
     EF_LOONGARCH_SINGLE_FLOAT   => 0x00000002,
     EF_LOONGARCH_DOUBLE_FLOAT   => 0x00000003,
@@ -184,6 +192,7 @@ my %elf_flags_mask = (
     EM_LOONGARCH()          => EF_LOONGARCH_ABI_MASK,
     EM_MIPS()               => EF_MIPS_ABI_MASK | EF_MIPS_ABI2,
     EM_PPC64()              => EF_PPC64_ABI64,
+    EM_RISCV()              => EF_RISCV_FLOAT_ABI_MASK | EF_RISCV_RVE,
 );
 
 sub get_format {
diff --git a/scripts/t/Dpkg_Arch.t b/scripts/t/Dpkg_Arch.t
index 95e6adf52..d2018593d 100644
--- a/scripts/t/Dpkg_Arch.t
+++ b/scripts/t/Dpkg_Arch.t
@@ -16,7 +16,7 @@
 use strict;
 use warnings;
 
-use Test::More tests => 18900;
+use Test::More tests => 19393;
 
 use_ok('Dpkg::Arch', qw(debarch_to_debtuple debarch_to_multiarch
                         debarch_eq debarch_is debarch_is_wildcard
@@ -28,7 +28,7 @@ use_ok('Dpkg::Arch', qw(debarch_to_debtuple debarch_to_multiarch
                         get_host_gnu_type
                         get_valid_arches));
 
-my $KNOWN_ARCHES_TOTAL = 569;
+my $KNOWN_ARCHES_TOTAL = 584;
 my @valid_arches = get_valid_arches();
 
 sub get_valid_wildcards
-- 
2.40.0

